README.rst
Netresearch distributed lock manager
====================================

.. contents:: Contents


Overview
========

NR_Lock is a distributed lock manager (DLM).
This is designed for multiple TYPO3 frontend servers using a single typo3temp
share and Database.
NR_Lock currently supports single Redis or Couchbase (experimental) caching
server as locking instance.
It replaces/extends the TYPO3 lock facility t3lib_lock.


Requirements
============

- Redis: => 2.4 - 2.8
- Couchbase: => ???
- PHP => 5.3 - 5.5
- TYPO3 => 4.5 - 6.2
- PHP/redis => 1.6


Installation
============

Installation is done by TYPO3 extension manager.


Configuration
=============

To set up NR_Lock as a replacement for t3lib_lock you need to alter your
localconf.php in /typo3conf/ and add the following lines::

    $arLockCfg = &$TYPO3_CONF_VARS['SYS']['locking'];

    if (extension_loaded('redis')) {
        $arLockCfg = array(
            'backend' => '\Netresearch\Lock\Backend_Redis',
            'options' => array(
                //'hostname'         => 'localhost',
                //'port'             => 6379,
                //'database'         => 0,
                //'password'         => '',
                //'compression'      => false,
                //'compressionLevel' => 1,
            ),
        );
    } elseif (extension_loaded('couchbase')) {
        $arLockCfg = array(
            'backend' => '\Netresearch\Lock\Backend_Couchbase',
            'options' => array(
                'servers' => array(
                    'localhost',
                ),
            ),
        );
    }


ToDo
====

- Support multiple caching server.
- Test using LIST for locks with redis (performance).
- Use nr_lock in TYPO3 6.2 ClassLoader - this is currently hard coded in the TYPO3 core

References
==========

- http://www.couchbase.com/
- http://www.redis.io/
- http://redis.io/topics/distlock
- https://engineering.gosquared.com/distributed-locks-using-redis
- https://github.com/ronnylt/redlock-php


ChangeLog
2014-08-25  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* Version: 2.0.0

2014-07-31  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* Version: 1.0.1

2014-07-31  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* Version: 1.0.0

2014-07-31  Sebastian Mendel  <sebastian.mendel@netresearch.de>

	* Initial commit



src/Lock.php
<?php
declare(encoding = 'UTF-8');
/**
 * Holds class definition of \Netresearch\Lock\Lock
 *
 * @category   Netresearch
 * @package    Lock
 * @subpackage Controller
 * @author     Sebastian Mendel <sebastian.mendel@netresearch.de>
 * @license    AGPL http://www.netresearch.de/
 * @link       http://www.netresearch.de/
 */

namespace Netresearch\Lock;

/**
 * Distributed lock manager, extended \t3lib_lock.
 *
 * UX class for t3lib_lock to handle locking in memory based backend.
 *
 * @category   Controller
 * @package    Netresearch
 * @subpackage Lock
 * @author     Sebastian Mendel <sebastian.mendel@netresearch.de>
 * @license    AGPL http://www.netresearch.de/
 * @link       http://www.netresearch.de/
 */
class Lock extends \t3lib_lock
{
    /**
     * @var Backend_Abstract Locking backend object
     */
    var $backend = null;

    /**
     * @var bool Flag indicating whether use TYPO3 original/classic locking
     */
    protected $bClassicLocking = true;

    /**
     * @var string Lock token
     */
    protected $strToken = null;



    /**
     * Returns new \t3lib_lock child class object.
     *
     * initializes locking, check input parameters and set variables accordingly.
     *
     * @param string  $id     ID to identify this lock in the system
     * @param string  $method Define which locking method to use. Defaults to "simple".
     * @param integer $loops  Number of times a locked resource is tried to be
     *                        acquired. Only used in manual locks method "simple".
     * @param integer $step   Milliseconds after lock acquire is retried.
     *                        $loops * $step results in the maximum delay of a
     *                        lock. Only used in manual lock method "simple".
     */
    public function __construct($id, $method = 'simple', $loops = 0, $step = 0)
    {
        $this->id = $id;

        if (intval($loops)) {
            $this->loops = intval($loops);
        }
        if (intval($step)) {
            $this->step = intval($step);
        }

        if ($this->initBackend()) {
            $this->bClassicLocking = false;
        } else {
            parent::__construct($id, $method, $loops, $step);
        }
    }



    /**
     * Initialize locking backend.
     *
     * @return bool success
     */
    protected function initBackend()
    {
        $arLockCfg = $GLOBALS['TYPO3_CONF_VARS']['SYS']['locking'];

        if (empty($arLockCfg['backend']) || ! class_exists($arLockCfg['backend'])) {
            \t3lib_div::devLog(
                'Locking backend class not found: "' . $arLockCfg['backend'] . '"',
                'nr_lock',
                \t3lib_div::SYSLOG_SEVERITY_ERROR
            );
            return false;
        }

        $this->backend = new $arLockCfg['backend']($arLockCfg['options']);

        return true;
    }



    /**
     * Acquire a lock and return when successful. If the lock is already open,
     * the client will retry {$this->loops} times with a {$this->step} delay.
     *
     * Returns true if the lock could be acquired immediately.
     * Returns false if the client had to wait for a former lock to be released.
     * Throws RuntimeException if no lock could be acquired.
     *
     * @return boolean TRUE if lock could be acquired without waiting, FALSE otherwise.
     * @throws \RuntimeException
     */
    public function acquire()
    {
        if ($this->bClassicLocking) {
            return parent::acquire();
        }

        try {
            $this->strToken = uniqid();
            $nTtl           = Lock::getLockTime();

            $arResult = $this->backend->acquire(
                $this->id, $nTtl, $this->strToken, $this->loops, $this->step
            );

            $this->isAcquired = true;

            if ($arResult['nWaitTime'] === 0) {
                return true;
            } else {
                return false;
            }
        } catch (\Exception $e) {
            throw new \RuntimeException('Lock could not be created', 0, $e);
        }
    }



    /**
     * Release the lock.
     *
     * @return boolean Returns TRUE on success or FALSE on failure
     */
    public function release()
    {
        if ($this->bClassicLocking) {
            return parent::release();
        }

        if (!$this->isAcquired) {
            return true;
        }

        $this->isAcquired = false;
        return $this->backend->release($this->id, $this->strToken);
    }



    /**
     * Returns default locking time.
     *
     * @return int
     */
    public static function getLockTime()
    {
        $nTtl = intval(ini_get('max_execution_time'));
        if ($nTtl <= 0) {
            $nTtl = 30;
        }

        return $nTtl;
    }
}

?>


ext_emconf.php
<?php

/***************************************************************
 * Extension Manager/Repository config file for ext "nr_lock".
 *
 * Auto generated 25-08-2014 14:17
 *
 * Manual updates:
 * Only the data in the array - everything else is removed by next
 * writing. "version" and "dependencies" must not be touched!
 ***************************************************************/

$EM_CONF[$_EXTKEY] = array(
	'title' => 'Netresearch distributed lock manager (DLM).',
	'description' => 'Provides a distributed lock manager (DLM) with support for Redis and Couchbase.',
	'category' => 'fe',
	'author' => 'Sebastian Mendel',
	'author_company' => 'Netresearch GmbH & Co.KG',
	'author_email' => 'sebastian.mendel@netresearch.de',
	'constraints' => array(
		'depends' => array(
			'php' => '5.3.0-5.5.99',
			'typo3' => '4.5.0-6.2.99',
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'state' => 'stable',
	'version' => '2.0.0',
	'_md5_values_when_last_written' => 'a:16:{s:9:"build.xml";s:4:"3f95";s:9:"ChangeLog";s:4:"bea1";s:16:"ext_autoload.php";s:4:"8322";s:12:"ext_icon.gif";s:4:"a459";s:17:"ext_localconf.php";s:4:"aa1d";s:10:"README.rst";s:4:"bf9c";s:18:"doc/ide_helper.php";s:4:"02db";s:12:"src/Lock.php";s:4:"0c7f";s:24:"src/Backend/Abstract.php";s:4:"a243";s:25:"src/Backend/Couchbase.php";s:4:"f6a0";s:21:"src/Backend/Redis.php";s:4:"1b76";s:19:"tests/bootstrap.php";s:4:"755b";s:18:"tests/LockTest.php";s:4:"5501";s:17:"tests/phpunit.xml";s:4:"c77f";s:30:"tests/Backend/AbstractTest.php";s:4:"26eb";s:27:"tests/Backend/RedisTest.php";s:4:"486d";}',
	'suggests' => array(
	),
);

?>

ext_localconf.php
<?php
/**
 * Extension config script
 *
 * PHP version 5
 *
 * @category   Netresearch
 * @package    Lock
 * @subpackage Configuration
 * @author     Sebastian Mendel <sebastian.mendel@netresearch.de>
 * @license    AGPL http://www.netresearch.de/
 * @link       http://www.netresearch.de/
 */

defined('TYPO3_MODE') or die('Access denied.');

/** @var array[] $TYPO3_CONF_VARS */
global $TYPO3_CONF_VARS;

$arLockCfg = $TYPO3_CONF_VARS['SYS']['locking'];

if (! empty($arLockCfg) && ! class_exists('ux_t3lib_lock', false)) {
    // we need to manually include these classes because autoloading does not work at this stage
    require_once t3lib_extMgm::extPath('nr_lock') . 'src/Backend/Abstract.php';
    require_once t3lib_extMgm::extPath('nr_lock') . 'src/Backend/Redis.php';
    require_once t3lib_extMgm::extPath('nr_lock') . 'src/Backend/Couchbase.php';
    require_once t3lib_extMgm::extPath('nr_lock') . 'src/Lock.php';

    // for TYPO3 4.5 - 4.7
    // extend t3lib_lock (XCLASS)
    class ux_t3lib_lock extends \Netresearch\Lock\Lock {}


    // for TYPO3 6.2
    // extensions using old class names
    $TYPO3_CONF_VARS['SYS']['Objects']['t3lib_lock']
        ['className'] = 'Netresearch\Lock\Lock';

    // extensions using new class names
    $TYPO3_CONF_VARS['SYS']['Objects']['TYPO3\CMS\Core\Locking\Locker']
        ['className'] = 'Netresearch\Lock\Lock';
}

?>


src/Backend/Abstract.php
<?php
declare(encoding = 'UTF-8');
/**
 * A caching backend which stores cache entries.
 *
 * @category   Netresearch
 * @package    Lock
 * @subpackage Controller
 * @author     Sebastian Mendel <sebastian.mendel@netresearch.de>
 * @license    AGPL http://www.netresearch.de/
 * @link       http://www.netresearch.de/
 */

namespace Netresearch\Lock;

/**
 * Class Netresearch_Lock_Backend_Abstract
 *
 * @category   Netresearch
 * @package    Lock
 * @subpackage Backend
 * @author     Sebastian Mendel <sebastian.mendel@netresearch.de>
 * @license    AGPL http://www.netresearch.de/
 * @link       http://www.netresearch.de/
 */
abstract class Backend_Abstract
{
    /**
     * Set the lock with the unique $strIdentifier and $strToken.
     *
     * @param string  $strIdentifier Lock unique identifier
     * @param string  $strToken      Token
     * @param integer $nTtl          Time to live
     *
     * @return boolean Success
     * @throws \Exception
     */
    public abstract function lock($strIdentifier, $strToken = null, $nTtl = null);



    /**
     * Drop/delete the lock identified by $strIdentifier und verified with $strToken.
     *
     * @param string $strIdentifier Lock unique identifier
     * @param string $strToken      Token
     *
     * @return boolean Success
     * @throws \Exception
     */
    public abstract function unlock($strIdentifier, $strToken = null);



    /**
     * Acquire a lock and return when successful. If the lock is already open,
     * the client will retry $nRetries times with a delay of $nSleep.
     *
     * It is important to know that the lock will be acquired in any case,
     * even if the request was blocked first. Therefore, the lock needs to be
     * released in every situation.
     *
     * Returns array with acquired lock data
     *
     * array(
     *     'strToken'  => [string]
     *     'nTtl'      => [integer]
     *     'nWaitTime' => [integer]
     * )
     *
     * Token:    $strToken or generated token
     * TTL:      Lock life time in seconds
     * WaitTime: Time till lock could be acquired - 0 means lock could be acquired immediately
     *
     * @param string  $strIdentifier Lock unique identifier
     * @param integer $nTtl          Time to live
     * @param string  $strToken      Token
     * @param integer $nRetries      Retry count for acquiring the lock
     * @param integer $nSleep        Sleep/wait time between retries
     *
     * @throws \Exception
     * @return array Acquired lock data
     */
    public function acquire(
        $strIdentifier, $nTtl = null, $strToken = null, $nRetries = 150, $nSleep = 100
    ) {
        $arResult = array();

        $strIdentifier = 'lock::' . $strIdentifier;

        if (null === $strToken) {
            $arResult['strToken'] = uniqid();
        } else {
            $arResult['strToken'] = $strToken;
        }
        $nRetryCountDown = $nRetries;
        if (null === $nTtl) {
            $arResult['nTtl'] = Lock::getLockTime();
        } else {
            $arResult['nTtl'] = $nTtl;
        }

        do {
            if ($this->lock($strIdentifier, $arResult['strToken'], $arResult['nTtl'])) {
                // returns false if we had to wait for existing lock to release
                // otherwise true if locking was immediately
                $arResult['nWaitTime'] = ($nRetries - $nRetryCountDown) * $nSleep;

                \t3lib_div::devLog(
                    'Acquired lock "' . $strIdentifier . '" after '
                    . $arResult['nWaitTime'] . ' ms',
                    'nr_lock',
                    \t3lib_div::SYSLOG_SEVERITY_INFO
                );

                return $arResult;
            }

            // Wait a random delay before to retry
            $nLockDelayTime = mt_rand(floor($nSleep / 2), $nSleep);
            usleep($nLockDelayTime * 1000);

            $nRetryCountDown--;

        } while ($nRetryCountDown > 0);

        \t3lib_div::devLog(
            'Could not acquire lock "' . $strIdentifier . '" after '
            . $nRetries . ' retries within ' . ($nSleep * $nRetries) . ' ms',
            'nr_lock',
            \t3lib_div::SYSLOG_SEVERITY_ERROR
        );

        throw new \Exception('Could not acquire lock');
    }



    /**
     * Release the lock identified by $strIdentifier und verified with $strToken. .
     *
     * @param string $strIdentifier Lock unique identifier
     * @param string $strToken      Lock token
     *
     * @return boolean Returns TRUE on success or FALSE on failure
     */
    public function release($strIdentifier, $strToken = null)
    {
        $strIdentifier = 'lock::' . $strIdentifier;

        if ($this->unlock($strIdentifier, $strToken)) {
            \t3lib_div::devLog(
                'Released lock "' . $strIdentifier . '"',
                'nr_lock',
                \t3lib_div::SYSLOG_SEVERITY_INFO
            );
            return true;
        } else {
            \t3lib_div::devLog(
                'Could not release lock "' . $strIdentifier . '"',
                'nr_lock',
                \t3lib_div::SYSLOG_SEVERITY_ERROR
            );
            return false;
        }
    }
}
?>


src/Backend/Couchbase.php
<?php
declare(encoding = 'UTF-8');
/**
 * A locking backend which stores locks by using Couchbase.
 *
 * @category   Netresearch
 * @package    Lock
 * @subpackage Controller
 * @author     Sebastian Mendel <sebastian.mendel@netresearch.de>
 * @license    AGPL http://www.netresearch.de/
 * @link       http://www.netresearch.de/
 */

namespace Netresearch\Lock;

/**
 * Class Netresearch_Lock_Backend_Couchbase
 *
 * @category   Netresearch
 * @package    Lock
 * @subpackage Backend
 * @author     Sebastian Mendel <sebastian.mendel@netresearch.de>
 * @license    AGPL http://www.netresearch.de/
 * @link       http://www.netresearch.de/
 */
class Backend_Couchbase extends Backend_Abstract
{
    const PREFIX = 'lock::';

    /**
     * @var \Couchbase
     */
    protected $couchbase = null;

    /**
     * @var string
     */
    protected $strCas = '';

    /**
     * Constructs this backend
     *
     * Options
     * <code>
     * array(
     *     // name       => default
     *     'hostname'    => 'localhost',
     *     'username'    => '',
     *     'password'    => '',
     *     'bucket'      => 'default',
     *     'persistent'  => true,
     *     'compression' => false,
     * )
     * </code>
     *
     * @param array $arOptions Couchbase connection options
     */
    public function __construct(array $arOptions)
    {
        $arOptionsDefault = array(
            'hostname'    => 'localhost',
            'username'    => '',
            'password'    => '',
            'bucket'      => 'default',
            'persistent'  => true,
            'compression' => false,
        );

        $arOptions = array_merge($arOptionsDefault, $arOptions);

        try {
            $this->couchbase = new \Couchbase(
                array($arOptions['hostname']),
                $arOptions['username'],
                $arOptions['password'],
                $arOptions['bucket'],
                $arOptions['persistent']
            );
        } catch (\Exception $e) {
            \t3lib_div::devLog(
                'Could not connect to couchbase: ' . $e->getMessage(),
                'nr_lock',
                \t3lib_div::SYSLOG_SEVERITY_ERROR
            );
            throw new \RuntimeException(
                'NR Lock could not connect to couchbase', 0, $e
            );
        }

        if ($arOptions['compression']) {
            $this->couchbase->setOption(
                COUCHBASE_OPT_COMPRESSION, COUCHBASE_COMPRESSION_FASTLZ
            );
        } else {
            $this->couchbase->setOption(
                COUCHBASE_OPT_COMPRESSION, COUCHBASE_COMPRESSION_NONE
            );
        }
    }



    /**
     * Get Lock.
     * Stores an exclusive key-value pair on couchbase instance.
     *
     * @param string  $strIdentifier Lock name/id - name/key of stored value
     * @param string  $strToken      Lock token - stored as value
     * @param integer $nTtl          Lock time to live in seconds
     *
     * @return boolean Success
     */
    public function lock($strIdentifier, $strToken = '', $nTtl = null)
    {
        if (empty($nTtl)) {
            $nTtl = Lock::getLockTime();
        }

        // set
        $strCas = $this->couchbase->set(
            self::PREFIX . $strIdentifier, $strToken, $nTtl
        );
        if (! $strCas) {
            return false;
        };

        // (get and) lock
        $result = $this->couchbase->getAndLock(
            self::PREFIX . $strIdentifier, $this->strCas, $nTtl
        );

        if ($result !== $strToken) {
            return false;
        }

        return true;
    }



    /**
     * Deletes lock from server.
     *
     * @param string $strIdentifier Lock name - key to access storage entry
     * @param string $strToken      Lock token - compared against stored value
     *
     * @return boolean Success
     */
    public function unlock($strIdentifier, $strToken = null)
    {
        $strCas = $this->couchbase->delete($strIdentifier, $this->strCas);

        return is_string($strCas);
    }
}
?>


src/Backend/Redis.php
<?php
declare(encoding = 'UTF-8');
/**
 * A locking backend which stores locks by using Redis.
 *
 * @category   Netresearch
 * @package    Lock
 * @subpackage Controller
 * @author     Sebastian Mendel <sebastian.mendel@netresearch.de>
 * @license    AGPL http://www.netresearch.de/
 * @link       http://www.netresearch.de/
 */

namespace Netresearch\Lock;

/**
 * Class Netresearch_Lock_Backend_Redis
 *
 * @category   Netresearch
 * @package    Lock
 * @subpackage Backend
 * @author     Sebastian Mendel <sebastian.mendel@netresearch.de>
 * @license    AGPL http://www.netresearch.de/
 * @link       http://www.netresearch.de/
 */
class Backend_Redis extends Backend_Abstract
{
    /**
     * @var \Redis
     */
    protected $redis = null;



    /**
     * Constructor.
     * Init redis connection.
     *
     * Options
     * <code>
     * array(
     *     // name            => default
     *     'hostname'         => 'localhost',
     *     'port'             => 6379,
     *     'database'         => 0,
     *     'password'         => '',
     *     'compression'      => false,
     *     'compressionLevel' => 1,
     * )
     * </code>
     *
     * @param array $arOptions Redis connection options
     */
    public function __construct(array $arOptions)
    {
        $this->redis = new \Redis();

        $arOptionsDefault = array(
            'hostname'         => 'localhost',
            'port'             => 6379,
            'database'         => 0,
            'password'         => '',
            'compression'      => false,
            'compressionLevel' => 1,
        );

        $arOptions = array_merge($arOptionsDefault, $arOptions);

        try {
            $bConnect = $this->redis->connect($arOptions['hostname'], $arOptions['port']);
        } catch (\Exception $e) {
            $bConnect = false;
        }

        if (false === $bConnect) {
            throw new \RuntimeException(
                'Could not connect to redis server.',
                0,
                empty($e) ? null : $e
            );
        }

        if (strlen($arOptions['password'])) {
            $success = $this->redis->auth($arOptions['password']);
            if (!$success) {
                throw new \RuntimeException(
                    'The given password was not accepted by the redis server.'
                );
            }
        }

        if ($arOptions['database'] > 0) {
            $success = $this->redis->select($arOptions['database']);
            if (!$success) {
                throw new \RuntimeException(
                    'The given database "' . $arOptions['database'] . '" could not be selected.'
                );
            }
        }
    }



    /**
     * Get Lock.
     * Stores an exclusive key-value pair on redis instance.
     *
     * @param string  $strIdentifier Lock name/id - name/key of stored value
     * @param string  $strToken      Lock token - stored as value
     * @param integer $nTtl          Lock time to live in seconds
     *
     * @return boolean Success
     */
    public function lock($strIdentifier, $strToken = '', $nTtl = null)
    {
        if (empty($nTtl)) {
            $nTtl = Lock::getLockTime();
        }

        $strLockScript = <<<LUA
--
-- Set a lock
--
-- KEYS[1]  - key
-- KEYS[2]  - ttl in ms
-- KEYS[3]  - token
local key   = KEYS[1]
local ttl   = KEYS[2]
local token = KEYS[3]

local lockSet = redis.call('setnx', key, token)

if lockSet == 1 then
  redis.call('pexpire', key, ttl)
end

return lockSet
LUA;

        // FIXME: better use SCRIPT LOAD und EVALSHA ?
        //$sha = $this->redis->script('load', $strLockScript);
        //$this->redis->evalSha($sha);

        // FIXME: Redis => 1.6
        //$this->redis->set($strIdentifier, $strToken, ['NX', 'PX' => $nTtl]);
        // or ->setnx() and ->setex()

        return (bool) $this->redis->eval(
            $strLockScript,
            array(
                $strIdentifier,
                $nTtl * 1000,
                $strToken,
            ),
            3
        );
    }



    /**
     * Release a lock.
     * Removes key-value from redis server - if token fits
     *
     * @param string $strIdentifier Lock name - key to access storage entry
     * @param string $strToken      Lock token - compared against stored value
     *
     * @return boolean Success
     */
    public function unlock($strIdentifier, $strToken = null)
    {
        if (null === $strToken) {
            $strUnlockScript
                = <<<LUA
--
-- Release a lock
--
-- KEYS[1]   - key
local key   = KEYS[1]

return redis.call("DEL", key)
LUA;
        } else {
            $strUnlockScript
                = <<<LUA
--
-- Release a lock
--
-- KEYS[1]   - key
-- ARGV[1]   - token (lock content)
local key   = KEYS[1]
local token = ARGV[1]

if redis.call("GET", key) == token then
    return redis.call("DEL", key)
else
    return 0
end
LUA;
        }

        return (bool) $this->redis->eval(
            $strUnlockScript, array($strIdentifier, $strToken), 1
        );
    }
}
?>
